scipy/stats/_stats_py.py

```diff
@@ -9018,64 +9029,136 @@ def confidence_interval(self, confidence_level=0.95):
             raise ValueError(message)
 
-        low_index = np.nan
-        high_index = np.nan
-
-        if alternative == 'less':
+        if n == 0:
+            zeros = np.zeros(p.shape, dtype=x.dtype)
+            low, high = zeros, zeros
+        elif alternative == 'less':
             p = 1 - confidence_level
-            low = -np.inf
-            high_index = int(bd.isf(p))
-            high = x[high_index] if high_index < n else np.nan
+            low = np.full(shape, -np.inf)
+            high_index = bd.isf(p).astype(int)
+            valid_index = high_index < n
+            high_index[~valid_index] = 0
+            high = np.where(valid_index, np.take_along_axis(x, high_index), np.nan)
         elif alternative == 'greater':
             p = 1 - confidence_level
-            low_index = int(bd.ppf(p)) - 1
-            low = x[low_index] if low_index >= 0 else np.nan
-            high = np.inf
+            low_index = (bd.ppf(p)).astype(int) - 1
+            valid_index = low_index >= 0
+            low_index[~valid_index] = 0
+            low = np.where(valid_index, np.take_along_axis(x, low_index), np.nan)
+            high = np.full(shape, np.inf)
         elif alternative == 'two-sided':
             p = (1 - confidence_level) / 2
-            low_index = int(bd.ppf(p)) - 1
-            low = x[low_index] if low_index >= 0 else np.nan
-            high_index = int(bd.isf(p))
-            high = x[high_index] if high_index < n else np.nan
-
+            low_index = (bd.ppf(p)).astype(int) - 1
+            valid_index = low_index >= 0
+            low_index[~valid_index] = 0
+            low = np.where(valid_index, np.take_along_axis(x, low_index), np.nan)
+            high_index = bd.isf(p).astype(int)
+            valid_index = high_index < n
+            high_index[~valid_index] = 0
+            high = np.where(valid_index, np.take_along_axis(x, high_index), np.nan)
+
+        args = self._axis, self._axis_none, self._keepdims, self._ndim, self._nan_out
+        low = _quantile_test_postprocess(low, *args)
+        high = _quantile_test_postprocess(high, *args)
         return ConfidenceInterval(low, high)
 
 
-def quantile_test_iv(x, q, p, alternative):
+def quantile_test_iv(x, q, p, alternative, axis, keepdims):
 
     x = np.atleast_1d(x)
-    message = '`x` must be a one-dimensional array of numbers.'
-    if x.ndim != 1 or not np.issubdtype(x.dtype, np.number):
+    message = '`x` must be an array of numbers.'
+    if not np.issubdtype(x.dtype, np.number):
         raise ValueError(message)
 
     q = np.array(q)[()]
-    message = "`q` must be a scalar."
-    if q.ndim != 0 or not np.issubdtype(q.dtype, np.number):
+    message = "`q` must be a scalar or array of numbers."
+    if not np.issubdtype(q.dtype, np.number):
         raise ValueError(message)
 
     p = np.array(p)[()]
-    message = "`p` must be a float strictly between 0 and 1."
-    if p.ndim != 0 or p >= 1 or p <= 0:
```

Comment:
Where `p >= 1` or `p <= 0`, results get replaced with NaN rather than raising. That way, a single bad entry doesn't cause the function to error out when there are a lot of valid entries to be processed.

Should add a test for these.